web/functions/index.js

```diff
@@ -0,0 +1,885 @@
+        var categoryId = "";
+        var categoryDoc = "";
+
+        //if this category collection or the specific category doesn't exist
+        if(categoryExist.empty) {
+
+            //make a new category
+            const newCategory = {
+                "name": `${req.body.category}`
+            }
+    
+            //add it to the category table
+            categoryDoc = await userRef.collection(categoryCollection).doc().set(newCategory);
+            categoryId = categoryDoc.id;
+        }
+        else {
+            var category = categoryExist.docs.map(category => {return category.id});
+            categoryId = category[0];
+        }
+
+        //make price string into an integer
+        const price = parseInt(req.body.price);
+
+        //object for a new one-off
+        const newOneOff = {
+            "name": `${req.body.name}`,
+            "categoryId": `${categoryId}`,
+            "color": `${req.body.color}`,
+            "price": `${price}`,
+            "date": `${req.body.date}`
+        }
+
+        const oneOffDoc = userRef.collection(oneOffCollection).doc();
+        const oneOffId = oneOffDoc.id;
+        await oneOffDoc.set(newOneOff);
+
+        const editedOneOff = {
+            "name": `${req.body.name}`,
+            "categoryId": `${categoryId}`,
+            "color": `${req.body.color}`,
+            "price": `${price}`,
+            "date": `${req.body.date}`,
+            "oneOffId": `${oneOffId}`
+        }
+
+        await userRef.collection(oneOffCollection).doc(`${oneOffId}`).update(editedOneOff);
+
+        res.status(201).send(`{
+            "userId": "${userId}",
+            "oneOffId": "${oneOffId}",
+            "name": "${req.body.name}",
+            "categoryId": "${categoryId}",
+            "color": "${req.body.color}",
+            "price": ${price},
+            "date": "${req.body.date}"
+        }`);
+
+    } catch (error) {
+        res.status(400).send(`${error.message}`)
+    }
+});
+
+//get one-offs
+app.post('/GetOneOffs', async (req, res) => {
+    res.set('Access-Control-Allow-Origin', '*');
+
+    try {
+
+        const userId = req.body.userId; 
+        const userRef = db.collection(userCollection).doc(`${userId}`); 
+
+         //check if the one-off already exists
+         const oneOffDocs = await userRef.collection(oneOffCollection).get();
+         if(!oneOffDocs.empty) {
+
+            var oneOffs = [];
+            oneOffDocs.docs.forEach(curOneOff => {
+                var oneOff = {
+                    "name":`${curOneOff.get("name")}`,
+                    "categoryId":`${curOneOff.get("categoryId")}`,
+                    "color":`${curOneOff.get("color")}`,
+                    "price":`${curOneOff.get("price")}`,
+                    "date":`${curOneOff.get("date")}`,
+                    "oneOffId":`${curOneOff.id}`,
+                }
+
+                oneOffs.push(oneOff);
+            })
```

Comment:
```suggestion
            const oneOffs = oneOffDocs.docs.map((curOneOff) => {
                return {
                    name: `${curOneOff.get('name')}`,
                    categoryId: `${curOneOff.get('categoryId')}`,
                    color: `${curOneOff.get('color')}`,
                    price: `${curOneOff.get('price')}`,
                    date: `${curOneOff.get('date')}`,
                    oneOffId: `${curOneOff.id}`,
                };
            });
```
This should def be a map